frontend/public/components/utils/resource-log.jsx

```diff
@@ -37,188 +32,202 @@ const LogControls = ({status, toggleStreaming, dropdown, onDownload}) => {
-    this._updateStatus = this._updateStatus.bind(this);
-    this._download = this._download.bind(this);
-
+    this._buffer = [];
     this.state = {
-      touched: Date.now(),
-      status: 'loading',
+      alive: true,
+      error: false,
+      lines: [],
+      linesBehind: 0,
+      stale: false,
+      status: 'loading'
     };
   }
 
   static getDerivedStateFromProps(nextProps, prevState) {
-    const newState = {};
-
-    if(nextProps.eof){
-      newState.status = prevState.status === 'paused' ? 'paused' : 'eof';
+    if (nextProps.alive !== prevState.alive) {
+      let newState = {};
+      newState.alive = nextProps.alive;
+      // Container changed from non-running to running state, so currently displayed logs are stale
+      if (nextProps.alive && !prevState.alive) {
+        newState.stale = true;
+      }
+      return newState;
     }
-    return newState;
+    return null;
   }
 
   componentDidMount() {
     super.componentDidMount();
-    this._beginStreaming();
+    this._wsInit(this.props);
   }
 
-  componentDidUpdate(prevProps){
-    // If container changed, restart streaming
-    if (this.props.containerName && this.props.containerName !== prevProps.containerName) {
-      this._restartStreaming();
+  componentDidUpdate(prevProps) {
+    // Container changed
+    if (prevProps.containerName !== this.props.containerName) {
+      this._restartStream();
     }
   }
 
   componentWillUnmount() {
     super.componentWillUnmount();
-    this._endStreaming();
+    this._wsDestroy();
   }
 
-  // Updates log status as long as eof has not been reached
-  _updateStatus(newStatus) {
-    // If log is at eof, don't update state to anything other than eof
-    this.setState((prevState, props) => {
-      return {
-        status: props.eof ? 'eof' : newStatus
-      };
-    });
+  // Download currently displayed log content
+  _download () {
+    const blob = new Blob([this._buffer.join('')], {type: 'text/plain;charset=utf-8'});
+    let filename = this.props.resourceName;
+    if (this.props.containerName) {
+      filename = `${filename}-${this.props.containerName}`;
+    }
+    saveAs(blob, `${filename}.log`);
   }
 
-  // use _updateStatus toggle streaming/paused state instead of directly
-  // setting logState
-  _toggleStreaming() {
-    this._updateStatus(this.state.status === 'streaming' ? 'paused' : 'streaming');
+  // Handler for websocket onclose event
+  _onClose(){
+    this.setState({ status: 'eof'});
   }
 
-  // separate function so that it can be throttled
-  _touch() {
+  // Handler for websocket onerror event
+  _onError(){
     this.setState({
-      touched: Date.now()
+      error: true
     });
   }
 
-  _pushToBuffer(data){
-    this._buffer.push(data);
-    this._touch();
-  }
-
-  _clearBuffer(){
-    this._buffer.clear();
-    this._touch();
+  // Handler for websocket onmessage event
+  _onMessage(msg){
+    const text = Base64.decode(msg);
```

Comment:
Can you verify that this handles unicode characters correctly?

Reply:
https://github.com/dankogai/js-base64
This is the library we are using, which does appear to handle unicode. This does remind me that I need to make sure this is defined as a dependency for us. It might just be there as a child dependency right now.

Reply:
why not atob/btoa?

Reply:
`atob` would not handle the full set of utf-8 characters. If there are any extended characters in the log content, it would break.